File: Algorithms/Sorting/Insertion Sort/insertion-sort.cpp
/*
 * Insertion Sort
 *
 */

#include <iostream>

using namespace std;

void printArray(int arr[], int size);

int insertionSort(int arr[], int n) {
    for (int i = 1; i < n; ++i) {
        int key = arr[i];
        int j = i-1;
        while (j >= 0 && arr[j] > key) {
            arr[j+1] = arr[j];
            j--;
        }
        arr[j+1] = key;
    }
    printArray(arr, n);
}

void printArray(int arr[], int size) {
    for (int i=0; i < size; i++)
        cout<<" "<<arr[i];
    cout<<endl;
}

int main(int argc, char const *argv[]) {
    int arr[] = {3,1,5,2,75,7,4,12,35,54,1,256,24,5736453,1,7};
    int n = sizeof(arr) / sizeof(arr[0]);
    cout<<"Insertion Sort => ";
    insertionSort(arr, n);
    return 0;
}

File: Algorithms/Sorting/Insertion Sort/insertion-sort.lua
--[[
    Insertion Sort
--]]

function insertion_sort(arr)
    for i=2,#arr do
        key = arr[i]
        j = i-1
        while j >= 0 and arr[j] > key do
            arr[j+1] = arr[j]
            j = j - 1
        end
        arr[j+1] = key
    end
end

arr = {1, 4, 2, 5, 1, 4}

print("Before Sorting")
for i = 1, #arr do
  io.write(arr[i], " ")
end
print()

insertion_sort(arr)

print("After Sorting")
for i = 1, #arr do
  io.write(arr[i], " ")
end
print()
